Bottom: cb5bb38f6bb0038b16a928ec5f6765a8e05b5dd3
Top:    ed212559043bacac232c481a125c8cfff34c5b04
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-04-06 00:05:27 +0100

GTK HID: Use clipping region when drawing DRC violation previews

Should speed up rendering quite a bit for non-trivial boards.


---

diff --git a/src/hid/gtk/gui-render-pixmap.c b/src/hid/gtk/gui-render-pixmap.c
index 525bde6..b24e6c9 100644
--- a/src/hid/gtk/gui-render-pixmap.c
+++ b/src/hid/gtk/gui-render-pixmap.c
@@ -49,6 +49,7 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   int save_left, save_top;
   int save_width, save_height;
   int save_view_width, save_view_height;
+  BoxType region;
 
   save_drawable = gport->drawable;
   save_zoom = gport->zoom;
@@ -79,7 +80,11 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   gdk_draw_rectangle (pixmap, gport->bg_gc, TRUE, 0, 0, width, height);
 
   /* call the drawing routine */
-  hid_expose_callback (&ghid_hid, NULL, NULL);
+  region.X1 = MIN(Px(0), Px(gport->width + 1));
+  region.Y1 = MIN(Py(0), Py(gport->height + 1));
+  region.X2 = MAX(Px(0), Px(gport->width + 1));
+  region.Y2 = MAX(Py(0), Py(gport->height + 1));
+  hid_expose_callback (&ghid_hid, &region, NULL);
 
   gport->drawable = save_drawable;
   gport->zoom = save_zoom;
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index 7a46672..1ed9082 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -532,9 +532,6 @@ void ghid_invalidate_all ();
 void ghid_get_coords (const char *msg, int *x, int *y);
 gint PCBChanged (int argc, char **argv, int x, int y);
 
-
-
-
 extern GdkPixmap *XC_hand_source, *XC_hand_mask;
 extern GdkPixmap *XC_lock_source, *XC_lock_mask;
 extern GdkPixmap *XC_clock_source, *XC_clock_mask;
